src/MongoDB.Driver/Encryption/ExplicitEncryptionLibMongoCryptController.cs

```diff
@@ -219,26 +218,25 @@ public async Task<BsonValue> DecryptFieldAsync(BsonBinaryData wrappedBinaryValue
         }
 
         // private methods
-        private IKmsKeyId GetKmsKeyId(string kmsProvider, IReadOnlyList<string> alternateKeyNames, BsonDocument masterKey)
+        private KmsKeyId GetKmsKeyId(string kmsProvider, IReadOnlyList<string> alternateKeyNames, BsonDocument masterKey)
         {
             IEnumerable<byte[]> wrappedAlternateKeyNamesBytes = null;
             if (alternateKeyNames != null)
             {
                 wrappedAlternateKeyNamesBytes = alternateKeyNames.Select(GetWrappedAlternateKeyNameBytes);
             }
 
-            switch (kmsProvider)
+            if (!Enum.TryParse<KmsType>(kmsProvider, true, out var kmsType))
             {
-                case "aws":
-                    var customerMasterKey = masterKey["key"].ToString();
-                    var endpoint = masterKey.GetValue("endpoint", null)?.ToString();
-                    var region = masterKey["region"].ToString();
-                    return new AwsKeyId(customerMasterKey, region, wrappedAlternateKeyNamesBytes, endpoint);
-                case "local":
-                    return wrappedAlternateKeyNamesBytes != null ? new LocalKeyId(wrappedAlternateKeyNamesBytes) : new LocalKeyId();
-                default:
-                    throw new ArgumentException($"Invalid kmsProvider {kmsProvider}.");
+                var message = $"The provided kms provider {kmsProvider} is not supported. The list of supported providers: {string.Join(", ", Enum.GetValues(typeof(KmsType)))}.";
+                throw new NotSupportedException(message);
             }
+            var dataKeyDocument = new BsonDocument("provider", kmsProvider.ToLower());
+            if (masterKey != null)
+            {
+                dataKeyDocument.AddRange(masterKey.Elements);
+            }
+            return new KmsKeyId(kmsType, dataKeyDocument.ToBson(), wrappedAlternateKeyNamesBytes);
```

Comment:
kmsType argument is not required here. The only reason why I left it as is improving developer experience. To avoid the need to convert bytes into something understandable mostly in debugging